Refactor global components and initialization process

mullefel opened the pull request:
Further refactoring for the initialization of components in the enclave-runtime. More re-use of existing/instantiated components instead of creating new instances.
The main idea was to extract the component initialization happens explicitly, in functions that have corresponding names, not implicitly. 

Explicit component initialization e-calls were added:
* Sidechain component initialization

Base components are initialized with the `init()` function. This includes all the components necessary for running the trusted and untrusted RPC servers in order to fix #600. In the course of doing that, I encountered another issue, #684, which I also had to fix in this PR.

Linked issues: #545 #600 #684
Closes #600 #684

mullefel commented on enclave-runtime/src/lib.rs:
```diff
@@ -383,57 +385,164 @@ pub unsafe extern "C" fn get_state(
-	server_addr_size: usize,
-) -> sgx_status_t {
-	let mut server_addr_encoded = slice::from_raw_parts(server_addr, server_addr_size);
+pub unsafe extern "C" fn init_enclave_base_components() -> sgx_status_t {
```
Initialize the 'base' components. Components that are used by both the parentchain and sidechain. The STF executor is a 'base' component at the moment, because indirect invocation immediately executes the indirect calls. In the future, it will be a sidechain only component.

mullefel commented on enclave-runtime/src/lib.rs:
```diff
@@ -383,57 +385,164 @@ pub unsafe extern "C" fn get_state(
+/// That means that this function will not return as long as the RPC server is running. The calling
+/// code should therefore spawn a new thread when calling this function.
+#[no_mangle]
+pub unsafe extern "C" fn init_direct_invocation_server(
+	server_addr: *const u8,
+	server_addr_size: usize,
+) -> sgx_status_t {
```
This function now only initializes the direct invocation server, as its name suggests. All the component initialization, that used to be here, is moved to the separate functions.

mullefel commented on enclave-runtime/src/top_pool_execution.rs:
```diff
@@ -152,41 +151,48 @@ fn execute_top_pool_trusted_calls_internal() -> Result<()> {
-		ExtrinsicsFactory::new(genesis_hash, authority.clone(), GLOBAL_NONCE_CACHE.clone());
+	let stf_executor = GLOBAL_STF_EXECUTOR_COMPONENT.get().ok_or_else(|| {
+		error!("Failed to retrieve stf executor component. Maybe it's not initialized?");
+		Error::ComponentNotInitialized
+	})?;
```
Get the STF executor from the global instance instead of creating another instance here.

haerdib commented on enclave-runtime/src/lib.rs:
```diff
@@ -383,57 +385,164 @@ pub unsafe extern "C" fn get_state(
+	sgx_status_t::SGX_SUCCESS
+}
+
+/// Initialize generic enclave components.
```
```suggestion
/// Initialize sidechain enclave components.
```

haerdib reviewed (changes requested):
This is certainly an improvement. However, I'm questioning the introduction of two separate E-Calls that need to be called in a specific order. Should this really be two separate E-calls, or rather one E-call that calls internal functions in a specific order?

I'm wondering what's the end goal here:
We now have several init E-calls, ranging from plain `init `https://github.com/integritee-network/worker/blob/73c8ddbef1e80adcba1e5b8ba28fd114a25509a5/enclave-runtime/src/lib.rs#L129
to new `components` inits and individual inits, such as the rpc-direct server https://github.com/integritee-network/worker/blob/73c8ddbef1e80adcba1e5b8ba28fd114a25509a5/enclave-runtime/src/lib.rs#L525

Do you have a roadmap in mind, or did you need to introduce this change to resolve something? If so, could you link this PR to a corresponding issue?

mullefel commented:
We already have multiple `init` functions, and they have to be called in the correct order, otherwise stuff breaks. 

We're experiencing some issues, where the order of starting up the direct invocation server and registering ourselves to the parentchain results in errors because of the initialization ordering. This PR is a step towards solving that (#600).
Of course, it would be nicer to just have one `init_components` function that initializes all the components. However, we have a couple of dependencies in the ordering. So we have base components that are needed for the parentchain registering to work, which is in turn needed to initialize more components. That resulted in this split of the initialization of components function. And for clarity, I think an `init_direct_server` should not instantiate components, that is just misleading.

clangenb commented on enclave-runtime/src/lib.rs:
```diff
@@ -383,57 +385,165 @@ pub unsafe extern "C" fn get_state(
 
-	let state_handler = Arc::new(GlobalFileStateHandler);
-	let ocall_api = Arc::new(OcallApi);
+	let stf_executor =
+		match GLOBAL_STF_EXECUTOR_COMPONENT.get() {
+			Some(a) => a,
+			None => {
+				error!("Failed to retrieve global STF executor component (maybe it is not initialized?)");
+				return sgx_status_t::SGX_ERROR_UNEXPECTED
+			},
+		};
```
We could have some way more ergonomic code if we wrap this code into an internal function and introduce a custom error. Snippet:

```
fn init_enclave_sidechain_components_internal() -> EnclaveResult {
...
let state_handler = GLOBAL_STATE_HANDLER_COMPONENT
                          .get().ok_or_else(|| EnclaveError::Other("Could not get ...")?;
}
```

Have you considered this?

mullefel replied:
Yes, I agree. I think I'll do a bit more refactoring to extract the initialization method into a separate module and move the initialization logic into there, with `Result` signatures 👍

clangenb reviewed (commented):
Nice, looks like yet another cleanup. Only Some minor comments.

mullefel commented:
I now moved all the component initialization into a separate module, `initialization.rs`, out of the `lib.rs`. Also, I removed the `init_enclave_base_components()` function again and moved everything into the regular `init()` function. 

In order to fix #600, I had to move the RPC server starting before the parentchain registering. Doing that revealed another issue, #684 - which I also fixed now in this PR.

haerdib commented on service/src/worker.rs:
```diff
@@ -32,32 +32,31 @@ use jsonrpsee::{
-		Self { _config, node_api, _enclave_api, peers }
-	}
-
-	// will soon be used.
```
🎉

haerdib commented on sidechain/consensus/aura/src/verifier.rs:
```diff
@@ -81,13 +81,8 @@ where
-		)?;
-
+		// We check the ancestry before the author, because the parentchain block might not match
+		// in case we try to import a sidechain block that was already imported longer than 1 parentchain block ago.
```
This comment is not quite self-explanatory if one does not know the background of it.

Suggestion:
`We need to check the ancestry first to ensure that an already imported block does not result in an author verification error.`

mullefel replied:
Yes, you're right. I'll update the comment 👍

haerdib reviewed (approved):
LGTM, Massive clean up, thanks! Only one minor comment

clangenb reviewed (commented):
It looks like a very nice clean up! I have only two comments.

mullefel commented on sidechain/consensus/common/src/block_import.rs:
```diff
@@ -111,10 +112,14 @@ where
-			self.peek_parentchain_header(&sidechain_block, parentchain_header)?;
+			self.peek_parentchain_header(&sidechain_block, parentchain_header)
+				.unwrap_or_else(|e| {
+					warn!("Could not peek parentchain block, returning latest parentchain block ({:?})", e);
```
It means that the sidechain block we're trying to import references a parentchain block that is older than what we currently have. This should only happen when onboarding, otherwise it's an indication of something going wrong - thus the `warn`.

mullefel commented on enclave-runtime/src/initialization.rs:
```diff
@@ -0,0 +1,264 @@
+}
+
+pub(crate) fn init_enclave_sidechain_components() -> EnclaveResult<()> {
+	let stf_executor = GLOBAL_STF_EXECUTOR_COMPONENT.get().ok_or_else(|| {
+		error!("Failed to retrieve global STF executor component (maybe it is not initialized?)");
+		Error::ComponentNotInitialized
+	})?;
```
Yes, agreed, this can be simplified and the amount of logging reduced 👍  thanks for pointing that out. Now I wish I had a `nameof()` function in Rust too, so I wouldn't need to type all the variable names as string for the errors 🙈

clangenb replied:
I agree to that. We could introduce an enum for the components and implement `toString` for that, but there might be a better way.

mullefel replied:
I think refactoring the `get` method of `ComponentContainer` might be sensible, to return an error that already contains all the information. And a container can also have a name that it puts into the error itself, so the caller does not have to care about the error contents.

mullefel replied:
The component container now knows its own name, and the getter returns a proper `Result` with an error that includes its name. That allowed me to reduce a lot of the getter calls to just `.get()?;`, which is certainly a lot cleaner. Thanks for insisting 😉

clangenb replied:
Coolio! Thanks for the extra effort. I think this was worth it!

clangenb reviewed (approved):
Looks good to me now! Thanks for the cleanup
